sources/web/datalab/polymer/modules/bigquery-file-manager/bigquery-file-manager.ts

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+class BigQueryMethodNotYetImplemented extends Error {
+  constructor(apiFunctionName: string) {
+    super('BigQuery method ' + apiFunctionName + ' is not implemented');
+  }
+}
+
+/**
+ * A file manager that wraps the BigQuery API so that we can browse BQ projects,
+ * datasets, and tables like a filesystem.
+ */
+class BigQueryFileManager implements FileManager {
+
+  /**
+   * Returns a DatalabFile object representing the file or directory requested
+   * @param path string path to requested file
+   * @param asText whether the file should be downloaded as plain text. This is
+   *               useful for downloading notebooks, which are by default read
+   *               as JSON, which doesn't preserve formatting.
+   */
+  public async get(path: string, asText?: boolean): Promise<DatalabFile> {
+    path; asText; // Make compiler happy until we implement this function.
+    throw new BigQueryMethodNotYetImplemented('get');
+  }
+
+  /**
+   * Uploads the given file object to the backend. The file's name, path, format,
+   * and content are required fields.
+   * @param model object containing file information to send to backend
+   */
+  public async save(file: DatalabFile): Promise<DatalabFile> {
+    file; // Make compiler happy until we implement this function.
+    throw new BigQueryMethodNotYetImplemented('save');
+  }
+
+  /**
+   * Returns a list of files at the target path, each implementing the
+   * DatalabFile interface. Two requests are made to /api/contents and
+   * /api/sessions to get this data.
+   * @param path current path to list files under
+   */
+  public list(path: string): Promise<DatalabFile[]> {
+    const pathParts = path.split('/').filter((part) => !!part);
```

Comment:
Please add a comment about why this works, and our assumption that slashes are never in any of the parts.